src/stm32f0xx/gpio.c

```diff
@@ -0,0 +1,76 @@
+#include "gpio.h"
+
+#include "stm32f0xx.h"
+#include "stm32f0xx_gpio.h"
+
+static GPIO_TypeDef *gpio_port_map[] = {GPIOA, GPIOB, GPIOC, GPIOD, GPIOE, GPIOF};
+
+// Map the uint8_t version of the pin to the uint16_t used by the gpio functions.
+// This isn't a linear map hence the switch case.
+static uint16_t prv_pin_map(uint8_t pin) {
+  assert_param(pin < 16);
+  return 0x01 << pin;
+}
+
+// Parse the GPIO settings and a pin number to create an initialization stuct to be used by the
+// CMSIS HAL.
+static GPIO_InitTypeDef *prv_parse_args(uint8_t pin, GPIOSettings *settings) {
+  GPIO_InitTypeDef init_struct;
+
+  // Parse the GPIOAltF settings which are used to modify the mode and Alt Function.
+  if (settings->alt_function == GPIO_ALTFN_ANALOG) {
+    init_struct.GPIO_Mode = GPIO_Mode_AN;
+  } else if (settings->alt_function == GPIO_ALTFN_NONE) {
+    init_struct.GPIO_Mode = settings->direction;
+  } else {
+    // Subtract 1 due to the offset of the enum from the ALTFN_NONE entry
+    init_struct.GPIO_Mode = GPIO_Mode_AF;
+  }
+  init_struct.GPIO_PuPd = settings->resistor;
+  init_struct.GPIO_Pin = prv_pin_map(pin);
+
+  // These are default values which are not intended to be changed.
+  init_struct.GPIO_Speed = GPIO_Speed_Level_1;
+  init_struct.GPIO_OType = GPIO_OType_PP;
+  return &init_struct;
+}
+
+void gpio_init() {
+  for (int i = 0; i < sizeof(gpio_port_map) / sizeof(GPIO_TypeDef); i++) {
+    // Sets the pin to a default reset mode.
+    // TODO(ckitagawa): determine if this is actually Lowest Power setting.
+    GPIO_DeInit(gpio_port_map[i]);
+  }
+}
+
+void gpio_init_pin(GPIOAddress *address, GPIOSettings *settings) {
+  GPIO_InitTypeDef *init_struct = prv_parse_args(address->pin, settings);
+  if (init_struct->GPIO_Mode == GPIO_Mode_AF) {
+    GPIO_PinAFConfig(gpio_port_map[address->port], init_struct->GPIO_Pin,
```

Comment:
You might as well move prv_parse_args into this function or make the function name more explicit.